docs/developers/talawa-api/error-handling.md

```diff
@@ -0,0 +1,1110 @@
+
+The interface `UserError` will act as an `interface contract`. The exact purpose for it will be cleared when we reach the  client side explaination part.
+```gql
+// now begins the Error unions types, we can consume many error types in this
+union SignUpError = EmailTaken | PasswordTooShort | UserError
+
+type EmailTaken implements UserError {
+  message: String!
+  path: String!
+  suggestion: String!
+}
+
+type PasswordTooShort implements UserError {
+  message: String!
+  path: String!
+  minimumLength: Int!
+}
+
+interface UserError {
+  message: String!
+  path: String!
+}
+```
+
+As you can see the `signUp` mutation has a return type of `SignUpResult!` which in turn contains `signUpData` with the return type `AuthData` and `signUpErrors` with return type of an array of the `SignUpError` union. This way the `signUp` mutation returns both the actual relevant data and the errors as result.
+
+```gql
+// Here is the return type of signup mutation  notice how the signUpData is nullable here, well that is optional.
+
+type SignUpResult {
+    signUpData : AuthData ,
+    signUpErrors : [SignUpError!]
+}
+
+type Mutation {
+   signUp(input: SignUpInput!): SignUpResult!
+}
+```
+
+Let us look at the pseudo code for the resolver now.
+
+```javascript
+const resolvers = {
+  Mutation: {
+    signUp: async (parent, args, context) => {
+          //the general approach of how this would work.
+           userObj = {} ,
+           signUpErrors = []
+           If(CHECK DUPLICATION OF EMAIL ) {
+                 KEEP THE EMAIL FIELD OF  USEROBJ TO BE RETURNED AS NULL ; 
+                 signUpErrors.push({
+                      __typename: "EmailTaken" ,
+                        message: "Email is already taken"
+                        path: "UserInput.email"
+                        suggestion: `Try to provide a unique mail or make sure you have not created an account 
+                        already`
+                 })
+           }
+
+           If (CHECK args.PASSWORD LENGTH) {
+                   signUpErrors.push({
+                      __typename: "PasswordTooShort" ,
+                        message: "Password length is too short"
+                        path: "UserInput.password"
+                        minimumLength: 8
+                   })
+           }
+           // Approach when we need to fall on a general error based on the interface contract `UserError`
+           If (CERTAIN CHECK WHERE WE WOULD NEED TO ADD THAT ERROR IN THE 
+                  signUpErrors ARRAY) {
+                    signUpErrors.push({
+                            __typename:"UserError" ,
+                            message: "message" , 
+                            path: "path"
+                    })
+                  
+            }
+
+          // Here we will be returning multiple errors in the form on an array of signUpErrors
+          if (signUpErrors) {
```

Comment:
Empty object(array) in javascript is considered a truthy value, this would trigger the if block it even though there might be no errors. Instead of this make the check on whether the array is empty or not. Similarly, in schema make the errors list non nullable, that way an array is always returned but client has to check if it is empty or not to determine whether an error occured.

Other choice is to only return an array for errors field when it has at least one item inside it, otherwise just return null. This wouldn't require changes to the schema.